mag/mag-als31300.c

```diff
@@ -0,0 +1,413 @@
+		log("failed to read register 0x%02x", addr);
+		return -1;
+	}
+
+	*valOut = (val[0] << 24) | (val[1] << 16) | (val[2] << 8) | val[3];
+	return 0;
+}
+
+
+static int getData(struct mag_data_s *data)
+{
+	int16_t rawX, rawY, rawZ, rawTemp;
+	uint32_t reg;
+
+	if (read32(DATA_REG1, &reg) < 0) {
+		return -1;
+	}
+
+	rawX = (reg >> 20) & 0x0FF0;
+	rawY = (reg >> 12) & 0x0FF0;
+	rawZ = (reg >> 4) & 0x0FF0;
+	rawTemp = (reg & 0x3F) << 6;
+
+	if (read32(DATA_REG2, &reg) < 0) {
+		return -1;
+	}
+
+	rawX |= ((reg >> 16) & 0x0F);
+	rawY |= ((reg >> 12) & 0x0F);
+	rawZ |= ((reg >> 8) & 0x0F);
+	rawTemp |= (reg & 0x3F);
+
+	rawX = (rawX ^ DATA_SIGN_BIT) - DATA_SIGN_BIT;
+	rawY = (rawY ^ DATA_SIGN_BIT) - DATA_SIGN_BIT;
+	rawZ = (rawZ ^ DATA_SIGN_BIT) - DATA_SIGN_BIT;
+
+	/* EM field in microTesla */
+	data->x = 1000 * (int32_t)rawX * DATA_GAUSS_MULTIPLIER_X / GAUSS_TO_MILITESLA_DIVIDER;
+	data->y = 1000 * (int32_t)rawY * DATA_GAUSS_MULTIPLIER_Y / GAUSS_TO_MILITESLA_DIVIDER;
+	data->z = 1000 * (int32_t)rawZ * DATA_GAUSS_MULTIPLIER_Z / GAUSS_TO_MILITESLA_DIVIDER;
+
+	/* temperature in miliCelsius */
+	data->temp = 1000 * DATA_TEMP_MULTIPLIER * (rawTemp - DATA_TEMP_SUBTRACTOR) / 4096;
+
+	return 0;
+}
+
+
+/* verify CFG1 contents to ensure we have ALS31300 under I2C_DEV_ADDR */
+static int verifyCfgReg(uint32_t val)
+{
+	if ((val & CFG_REG1_CUSTOMER_EE_MASK) != 0) {
+		return -1; /* we don't store anything in customer EEPROM - should be 0 */
+	}
+
+	if ((val & CFG_REG1_I2C_ADDR_MASK) != 0) {
+		return -1; /* factory configuration: eeprom-based i2c address should be 0x0 */
+	}
+
+	if (((val & CFG_REG1_ENABLE_X) == 0) || ((val & CFG_REG1_ENABLE_Y) == 0) || ((val & CFG_REG1_ENABLE_Z) == 0)) {
+		return -1; /* all channels should be enabled */
+	}
+
+	if ((val & CFG_REG1_I2C_THRESH) == 0) {
+		return -1; /* factory configuration: 1.8V i2c compatible mode enabled */
+	}
+
+	return 0;
+}
+
+
+/* probe for ALS31300 - return 0 if found, negative value otherwise */
+static int probe(void)
+{
+	uint32_t reg;
+	if (read32(CFG_REG1, &reg) < 0) {
+		/* no device with I2C_DEV_ADDR present on the i2c bus */
+		return -1;
+	}
+
+	if (verifyCfgReg(reg) < 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int handleRead(unsigned int devId, char *buf, size_t size)
+{
+	struct mag_data_s data;
+	int len;
+
+	if (getData(&data) < 0) {
+		return -EIO;
+	}
+	else {
+		switch (devId) {
+			case DEV_ID_X:
+				len = snprintf(buf, size, "%d\n", data.x);
+				return min(len, (int)size);
+			case DEV_ID_Y:
+				len = snprintf(buf, size, "%d\n", data.y);
+				return min(len, (int)size);
+			case DEV_ID_Z:
+				len = snprintf(buf, size, "%d\n", data.z);
+				return min(len, (int)size);
+			case DEV_ID_TEMP:
+				len = snprintf(buf, size, "%d\n", data.temp);
+				return min(len, (int)size);
+			default:
+				return -ENODEV;
+		}
+	}
+
+	return 0;
+}
+
+
+static void handleDumpMode(void)
+{
+	struct mag_data_s data;
+
+	while (1) {
```

Comment:
I don't know if you case about MISRA, but while (1) causes implicit essential type cast int -> bool
Examples of allowed infinite loops:
while (1 == 1)
while (true) (it's ok, I guess, but we tend to avoid stdbool)
for (;;) (advised)